### PR TITLE
FlyteContext Overhaul

### PR DESCRIPTION
**Pull request by kumare3**
Signed-off-by: Ketan Umare <ketan.umare@gmail.com>

# TL;DR
Flytekit Context is not a completely functional stack. Every stack element is immutable and the stack maintains its state with valid assertions in case of problems.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1001

### CODE REVIEW

File: flytekit/core/context_manager.py [hunk 1]
```diff
@@ -243,268 +194,353 @@ class Mode(Enum):
...
 
-    @property
-    def compilation_state(self) -> Optional[CompilationState]:
-        if self._compilation_state is not None:
-            return self._compilation_state
-        elif self._parent is not None:
-            return self._parent.compilation_state
-        else:
-            return None
+    def new_compilation_state(self, prefix: str = "") -> CompilationState:
```
Comment (wild-endeavor) [inline flytekit/core/context_manager.py:322]: I know i'm mostly to blame for this, but with this change, there's now
* `FlyteContext.with_compilation_state`
* `FlyteContext.with_new_compilation_state`
* `FlyteContext.new_compilation_state`
which are repeated for the Builder object
* `FlyteContext.Builder.with_compilation_state`
* `FlyteContext.Builder.with_new_compilation_state`
* `FlyteContext.Builder.new_compilation_state`

Can we at least get rid of `new_compilation_state`? It doesn't seem to do anything except call the constructor in both places.

File: flytekit/bin/entrypoint.py [hunk 2]
```diff
@@ -229,12 +235,20 @@ def setup_execution(
         # The reason we need this is because of dynamic tasks. Even if we move compilation all to Admin,
         # if a dynamic task calls some task, t1, we have to write to the DJ Spec the correct task
         # identifier for t1.
-        with ctx.new_serialization_settings(serialization_settings=serialization_settings) as ctx:
+        with FlyteContextManager.with_context(ctx.with_serialization_settings(serialization_settings).build()) as ctx:
             # Because execution states do not look up the context chain, it has to be made last
```
Comment (wild-endeavor) [inline flytekit/bin/entrypoint.py:239]: this comment isn't true anymore right?

File: flytekit/core/context_manager.py (hunk 1 above)
Comment (kumare3) [reply flytekit/core/context_manager.py:322]: We can, I was just so afraid to get rid of it

File: flytekit/core/node.py [hunk 3]
```diff
@@ -21,6 +21,8 @@ def __init__(
         upstream_nodes: List[Node],
         flyte_entity: Any,
     ):
+        if id is None:
```
Comment (kumare3) [inline flytekit/core/node.py:24]: @wild-endeavor this line seems odd?

**Review by wild-endeavor — APPROVED** [review]

File: flytekit/core/node.py (hunk 3 above)
Comment (wild-endeavor) [reply flytekit/core/node.py:24]: what do you mean?  like why would it ever be None?

**Comment by wild-endeavor** [comment]
✈️ 🙏

**Pull request merged**